cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -318,6 +319,32 @@ private void getScheduleStatus(HttpResponder responder, String appId, String nam
     }
   }
 
+  /**
+   * Stops the particular run of the Workflow or MapReduce program.
+   */
+  @POST
+  @Path("/apps/{app-id}/{type}/{id}/runs/{run-id}/stop")
+  public void performRunLevelStop(HttpRequest request, HttpResponder responder,
+                                  @PathParam("namespace-id") String namespaceId,
+                                  @PathParam("app-id") String appId,
+                                  @PathParam("type") String type,
+                                  @PathParam("id") String id,
+                                  @PathParam("run-id") String runId) {
+    try {
+      ProgramType programType = ProgramType.valueOfCategoryName(type);
+      if (!isRunLevelActionAllowed(programType)) {
+        responder.sendString(HttpResponseStatus.FORBIDDEN,
+                             "Run level operation is allowed only for MapReduce and Workflow program.");
+      }
+
+      Id.Program program = Id.Program.from(namespaceId, appId, programType, id);
+      AppFabricServiceStatus status = stop(program, runId);
+      responder.sendString(status.getCode(), status.getMessage());
+     } catch (IllegalArgumentException e) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, String.format("Invalid program type %s", type));
```

Comment:
Throw the `BadRequestException` instead. It's the new way to have one place to handle exceptions from handlers and responses properly.